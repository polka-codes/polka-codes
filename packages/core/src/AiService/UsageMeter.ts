/**
 * UsageMeter class for tracking API usage metrics
 * Generated by polka.codes
 */

import type { ApiUsage } from './AiServiceBase'
import { type ModelInfo, type ModelPricingConfig, applyCustomPricing } from './ModelInfo'

export class UsageMeter {
  #modelPricingConfig?: ModelPricingConfig
  #usage = {
    inputTokens: 0,
    outputTokens: 0,
    cacheWriteTokens: 0,
    cacheReadTokens: 0,
    totalCost: 0,
  }

  #messageCount = 0

  readonly maxCost: number
  readonly maxMessageCount: number

  constructor(
    options: {
      maxCost?: number
      maxMessageCount?: number
      modelPricingConfig?: ModelPricingConfig
    } = {},
  ) {
    // default to some something is definitely wrong if ever exceeded value
    this.maxCost = options.maxCost || 1000
    this.maxMessageCount = options.maxMessageCount || 1000
    this.#modelPricingConfig = options.modelPricingConfig
  }

  /**
   * Add usage metrics to the current totals
   */
  addUsage(usage: Partial<ApiUsage>, model?: ModelInfo, provider?: string, modelId?: string): void {
    // Use atomic operations for thread safety
    this.#usage.inputTokens += usage.inputTokens ?? 0
    this.#usage.outputTokens += usage.outputTokens ?? 0
    this.#usage.cacheWriteTokens += usage.cacheWriteTokens ?? 0
    this.#usage.cacheReadTokens += usage.cacheReadTokens ?? 0

    if (!usage.totalCost && model) {
      // Apply custom pricing if available
      const modelWithCustomPricing =
        provider && modelId && this.#modelPricingConfig ? applyCustomPricing(model, modelId, provider, this.#modelPricingConfig) : model

      usage.totalCost =
        ((modelWithCustomPricing.inputPrice ?? 0) * (usage.inputTokens ?? 0) +
          (modelWithCustomPricing.outputPrice ?? 0) * (usage.outputTokens ?? 0) +
          (modelWithCustomPricing.cacheWritesPrice ?? 0) * (usage.cacheWriteTokens ?? 0) +
          (modelWithCustomPricing.cacheReadsPrice ?? 0) * (usage.cacheReadTokens ?? 0)) /
        1_000_000
    }

    this.#usage.totalCost += usage.totalCost ?? 0
  }

  incrementMessageCount(count = 1) {
    this.#messageCount += count
  }

  isLimitExceeded() {
    const messageCount = this.#messageCount >= this.maxMessageCount
    const cost = this.#usage.totalCost >= this.maxCost
    return {
      messageCount,
      cost,
      result: messageCount || cost,
    }
  }

  /**
   * Get current usage totals
   */
  get usage(): ApiUsage {
    return { ...this.#usage }
  }

  printUsage() {
    const { inputTokens, outputTokens, cacheReadTokens, cacheWriteTokens } = this.#usage
    const allTokensZero = inputTokens === 0 && outputTokens === 0 && cacheReadTokens === 0 && cacheWriteTokens === 0

    console.log('Usages:')
    if (!allTokensZero) {
      console.log(`Input tokens: ${this.#usage.inputTokens}`)
      console.log(`Output tokens: ${this.#usage.outputTokens}`)
      console.log(`Cache read tokens: ${this.#usage.cacheReadTokens}`)
      console.log(`Cache write tokens: ${this.#usage.cacheWriteTokens}`)
    }
    console.log(`Total cost: ${this.#usage.totalCost}`)
  }
}
