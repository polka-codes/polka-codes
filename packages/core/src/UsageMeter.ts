/**
 * UsageMeter class for tracking API usage metrics
 * Generated by polka.codes
 */

import type { LanguageModelV2, LanguageModelV2Usage } from '@ai-sdk/provider'

export type ModelInfo = {
  inputPrice: number
  outputPrice: number
  cacheWritesPrice: number
  cacheReadsPrice: number
}

type Totals = { input: number; output: number; cachedRead: number; cost: number; messageCount: number }

/**
 * Tracks token / cost usage across any mix of LLM models.
 * Supports optional caps on total messages and total cost.
 */
export class UsageMeter {
  #totals: Totals = { input: 0, output: 0, cachedRead: 0, cost: 0, messageCount: 0 }

  readonly #modelInfos: Record<string, ModelInfo>
  readonly #maxMessages: number
  readonly #maxCost: number

  constructor(modelInfos: Record<string, Record<string, Partial<ModelInfo>>> = {}, opts: { maxMessages?: number; maxCost?: number } = {}) {
    const infos: Record<string, ModelInfo> = {}
    for (const [provider, providerInfo] of Object.entries(modelInfos)) {
      for (const [model, modelInfo] of Object.entries(providerInfo)) {
        // make google-vertex to google
        // and unify modelId
        infos[`${provider.split('-')[0]}:${model.replace(/[.-]/g, '')}`] = {
          inputPrice: modelInfo.inputPrice ?? 0,
          outputPrice: modelInfo.outputPrice ?? 0,
          cacheWritesPrice: modelInfo.cacheWritesPrice ?? 0,
          cacheReadsPrice: modelInfo.cacheReadsPrice ?? 0,
        }
      }
    }

    this.#modelInfos = infos
    this.#maxMessages = opts.maxMessages ?? 1000
    this.#maxCost = opts.maxCost ?? 100
  }

  #calculageUsage(usage: LanguageModelV2Usage, providerMetadata: any, modelInfo: ModelInfo) {
    const providerMetadataKey = Object.keys(providerMetadata ?? {})[0]
    const metadata = providerMetadata?.[providerMetadataKey] ?? {}

    switch (providerMetadataKey) {
      case 'openrouter':
        return {
          input: usage.inputTokens ?? 0,
          output: usage.outputTokens ?? 0,
          cachedRead: usage.cachedInputTokens ?? 0,
          cost: metadata.usage?.cost ?? 0,
        }
      case 'anthropic': {
        const cachedRead = usage.cachedInputTokens ?? 0
        const cacheWrite = metadata?.promptCacheMissTokens ?? 0
        const input = usage.inputTokens ?? 0
        const output = usage.outputTokens ?? 0

        return {
          input: input + cacheWrite + cachedRead,
          output,
          cachedRead,
          cost:
            (input * modelInfo.inputPrice +
              output * modelInfo.outputPrice +
              cacheWrite * modelInfo.cacheWritesPrice +
              cachedRead * modelInfo.cacheReadsPrice) /
            1_000_000,
        }
      }
      case 'deepseek': {
        const cachedRead = usage.cachedInputTokens ?? 0
        const cacheWrite = metadata.promptCacheMissTokens ?? 0
        const input = usage.inputTokens ?? 0
        const output = usage.outputTokens ?? 0

        return {
          input,
          output,
          cachedRead,
          cost: (output * modelInfo.outputPrice + cacheWrite * modelInfo.inputPrice + cachedRead * modelInfo.cacheReadsPrice) / 1_000_000,
        }
      }
      default: {
        const cachedRead = usage.cachedInputTokens ?? 0
        const input = usage.inputTokens ?? 0
        const output = usage.outputTokens ?? 0

        return {
          input,
          output,
          cachedRead,
          cost: (input * modelInfo.inputPrice + output * modelInfo.outputPrice) / 1_000_000,
        }
      }
    }
  }

  addUsage(
    llm: LanguageModelV2,
    resp: { usage: LanguageModelV2Usage; providerMetadata?: any } | { totalUsage: LanguageModelV2Usage; providerMetadata?: any },
    options: { modelInfo?: ModelInfo } = {},
  ) {
    const modelInfo = options.modelInfo ??
      // make google.vertex.chat to google
      // and anthropic.messages to anthropic
      this.#modelInfos[`${llm.provider.split('.')[0]}:${llm.modelId.replace(/[.-]/g, '')}`] ?? {
        inputPrice: 0,
        outputPrice: 0,
        cacheWritesPrice: 0,
        cacheReadsPrice: 0,
      }

    const usage = 'totalUsage' in resp ? resp.totalUsage : resp.usage
    const result = this.#calculageUsage(usage, resp.providerMetadata, modelInfo)

    this.#totals.input += result.input || 0
    this.#totals.output += result.output || 0
    this.#totals.cachedRead += result.cachedRead || 0
    this.#totals.cost += result.cost || 0
    this.#totals.messageCount += 1
  }

  /** Override the running totals (e.g., restore from saved state). */
  setUsage(newUsage: Partial<Totals>) {
    if (newUsage.input != null) this.#totals.input = newUsage.input
    if (newUsage.output != null) this.#totals.output = newUsage.output
    if (newUsage.cachedRead != null) this.#totals.cachedRead = newUsage.cachedRead
    if (newUsage.cost != null) this.#totals.cost = newUsage.cost
    if (newUsage.messageCount != null) this.#totals.messageCount = newUsage.messageCount
  }

  /** Manually bump the message count (useful if you record some messages without token info). */
  incrementMessageCount(n = 1) {
    this.#totals.messageCount += n
  }

  /** Reset the running totals. */
  resetUsage() {
    this.#totals = { input: 0, output: 0, cachedRead: 0, cost: 0, messageCount: 0 }
  }

  /** Return true once either messages or cost exceed the configured caps. */
  isLimitExceeded() {
    const messageCount = this.#maxMessages > 0 && this.#totals.messageCount >= this.#maxMessages
    const cost = this.#maxCost > 0 && this.#totals.cost >= this.#maxCost

    return {
      messageCount,
      maxMessages: this.#maxMessages,
      cost,
      maxCost: this.#maxCost,
      result: messageCount || cost,
    }
  }

  /** Same as isLimitExceeded but throws an error if a limit is hit. */
  checkLimit() {
    const result = this.isLimitExceeded()
    if (result.result) {
      throw new Error(
        `Usage limit exceeded. Message count: ${result.messageCount}/${result.maxMessages}, cost: ${result.cost}/${result.maxCost}`,
      )
    }
  }

  /** Getter for the aggregated totals (immutable copy). */
  get usage() {
    return { ...this.#totals }
  }

  /** Print a concise usage summary to console. */
  printUsage(customConsole: Console = console) {
    const u = this.usage
    customConsole.log(
      `Usage - messages: ${u.messageCount}, input: ${u.input}, cached: ${u.cachedRead}, ` +
        `output: ${u.output}, cost: $${u.cost.toFixed(4)}`,
    )
  }

  onFinishHandler(llm: LanguageModelV2) {
    return (evt: { totalUsage: LanguageModelV2Usage; providerMetadata: any }) => {
      this.addUsage(llm, evt)
    }
  }
}
