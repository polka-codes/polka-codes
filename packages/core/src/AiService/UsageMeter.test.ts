/**
 * Tests for UsageMeter class
 * Generated by polka.codes
 */

import { describe, expect, test } from 'bun:test'
import { UsageMeter } from './UsageMeter'

describe('UsageMeter', () => {
  test('should initialize with zero values', () => {
    const meter = new UsageMeter()
    expect(meter.usage).toMatchSnapshot()
  })

  test('should correctly add usage metrics', () => {
    const meter = new UsageMeter()

    meter.addUsage({
      inputTokens: 100,
      outputTokens: 50,
      cacheWriteTokens: 25,
      cacheReadTokens: 10,
      totalCost: 0.002,
    })

    expect(meter.usage).toMatchSnapshot()
  })

  test('should handle partial usage updates', () => {
    const meter = new UsageMeter()

    meter.addUsage({ inputTokens: 100 })
    meter.addUsage({ outputTokens: 50 })
    meter.addUsage({ totalCost: 0.001 })

    expect(meter.usage).toMatchSnapshot()
  })

  test('should handle multiple consecutive updates', () => {
    const meter = new UsageMeter()

    meter.addUsage({ inputTokens: 50, outputTokens: 25 })
    meter.addUsage({ inputTokens: 30, cacheReadTokens: 10 })
    meter.addUsage({ outputTokens: 15, totalCost: 0.001 })

    expect(meter.usage).toMatchSnapshot()
  })
})
