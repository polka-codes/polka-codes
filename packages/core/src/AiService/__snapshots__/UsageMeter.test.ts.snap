// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`UsageMeter should initialize with zero values 1`] = `
{
  "cacheReadTokens": 0,
  "cacheWriteTokens": 0,
  "inputTokens": 0,
  "messageCount": 0,
  "outputTokens": 0,
  "totalCost": 0,
}
`;

exports[`UsageMeter should correctly add usage metrics 1`] = `
{
  "cacheReadTokens": 10,
  "cacheWriteTokens": 25,
  "inputTokens": 100,
  "messageCount": 0,
  "outputTokens": 50,
  "totalCost": 0.002,
}
`;

exports[`UsageMeter should handle partial usage updates 1`] = `
{
  "cacheReadTokens": 0,
  "cacheWriteTokens": 0,
  "inputTokens": 100,
  "messageCount": 0,
  "outputTokens": 50,
  "totalCost": 0.001,
}
`;

exports[`UsageMeter should handle multiple consecutive updates 1`] = `
{
  "cacheReadTokens": 10,
  "cacheWriteTokens": 0,
  "inputTokens": 80,
  "messageCount": 0,
  "outputTokens": 40,
  "totalCost": 0.001,
}
`;

exports[`UsageMeter should calculate cost when not provided but model info is available 1`] = `
{
  "cacheReadTokens": 50,
  "cacheWriteTokens": 100,
  "inputTokens": 1000,
  "messageCount": 0,
  "outputTokens": 500,
  "totalCost": 0.0009065,
}
`;

exports[`UsageMeter should not calculate cost when model info is not available 1`] = `
{
  "cacheReadTokens": 0,
  "cacheWriteTokens": 0,
  "inputTokens": 1000,
  "messageCount": 0,
  "outputTokens": 500,
  "totalCost": 0,
}
`;

exports[`UsageMeter should use custom pricing configuration 1`] = `
{
  "cacheReadTokens": 100,
  "cacheWriteTokens": 200,
  "inputTokens": 1000,
  "messageCount": 0,
  "outputTokens": 500,
  "totalCost": 0.0105,
}
`;

exports[`UsageMeter should override existing pricing with custom configuration 1`] = `
{
  "cacheReadTokens": 100,
  "cacheWriteTokens": 200,
  "inputTokens": 1000,
  "messageCount": 0,
  "outputTokens": 500,
  "totalCost": 0.00211,
}
`;
